### PR TITLE
Make constraint optimizer pluggable (fixes #422)

### LINKED ISSUE #422: Hyperparameters for constrained optimization
Hey guys,
I was wondering if you might have some guidelines for selecting 
good hyperparameters for the constrained optimizer of the lossprogram
like c_warm_up_iterations, c_learning rate etc. 
Are the set standard values based on empirical evidence?

Also, is there a specific reason why Adam is specified as coptim over SGD?

### PR DESCRIPTION
**Pull request by Ashut0sh-mishra**
## Summary

Fixes #422.

Previously the constraint optimizer `self.copt` was hardcoded to
`torch.optim.Adam` in three places inside `domiknows/program/lossprogram.py`
(`PrimalDualProgram.train`, `InferenceProgram.train`, `SampleLossProgram.train`),
so users who wanted SGD (or any other optimizer / extra hyperparameters like
`momentum`, `weight_decay`, `betas`) had no clean way to swap it out — matching
the maintainer's suggestion on the issue that the user should be able to
choose the optimizer.

This PR makes the constraint optimizer fully pluggable, with zero breakage to
existing behaviour.

## What changed

**`domiknows/program/lossprogram.py`**
- `LossProgram.__init__` now accepts two new optional kwargs:
  - `copt_class` — the optimizer class (defaults to `torch.optim.Adam`, so
    current behaviour is preserved exactly).
  - `copt_kwargs` — a `dict` of extra kwargs forwarded to the optimizer
    (e.g. `{'momentum': 0.9}` for SGD, `{'weight_decay': 0.01}` for AdamW).
- New `LossProgram._make_copt(lr)` helper centralises optimizer
  construction and returns `None` when the cmodel has no parameters
  (the same guard that used to be duplicated in three places).
- `PrimalDualProgram.train`, `InferenceProgram.train` and
  `SampleLossProgram.train` no longer mention `torch.optim.Adam` — they all
  delegate to `self._make_copt(lr)`.
- Passing `'lr'` inside `copt_kwargs` is rejected early with a clear
  `ValueError`, because the learning rate must flow through `c_lr` in
  `train()` and silently dropping a duplicate would hide user intent.

## Usage

```python
from domiknows.program.lossprogram import InferenceProgram
import torch

# Default — identical to the previous behaviour (Adam):
prog = InferenceProgram(graph, Model)

# Swap to SGD with momentum:
prog = InferenceProgram(
    graph, Model,
    copt_class=torch.optim.SGD,
    copt_kwargs={'momentum': 0.9},
)
prog.train(data, c_lr=0.01, c_warmup_iters=20)

# AdamW with weight decay:
prog = InferenceProgram(
    graph, Model,
    copt_class=torch.optim.AdamW,
    copt_kwargs={'weight_decay': 0.1},
)

### CODE REVIEW

**Comment by Ashut0sh-mishra** [comment]
@kordjamshidi Hi, just checking in on this PR whenever you get a chance to review it.
Please let me know if any changes or updates are needed from my side. Thanks!
